src/recode.jl

```diff
@@ -57,13 +57,11 @@ function recode!(dest::AbstractArray{T}, src::AbstractArray, default::Any, pairs
         throw(DimensionMismatch("dest and src must be of the same length (got $(length(dest)) and $(length(src)))"))
     end
 
-    opt_pairs = map(optimize_pair, pairs)
-
     @inbounds for i in eachindex(dest, src)
         x = src[i]
 
-        for j in 1:length(opt_pairs)
-            p = opt_pairs[j]
+        for j in 1:length(pairs)
+            p = optimize_pair(pairs[j])
```

Comment:
This looks problematic when `pairs[j]` is an array as it will create a `Set` for each item. Have you investigated why `map(optimize_pair, pairs)` isn't inferred correctly? In theory it should, and I think I had carefully checked performance when I wrote that function so I'm surprised it's not the case.

Reply:
`pairs[j]` is never an array, it is always a `Pair`. I just replaced the map with a loop, but it still iterates over the exact same thing, so I don't think that should ever be problematic.

I was also surprised that `map(optimize_pair, pairs)` wasn't inferred, some change in the compiler must have triggered this. I stumbled into this easy fix, so then I didn't investigate further.

Reply:
I mean when the first element is an array to that the call hits `optimize_pair(pair::Pair{<:AbstractArray})`.

Reply:
The problem is that currently we create a `Set` once for each pair, and now we create it for each pair for each entry in the input array.

Reply:
Why is that the case?
With this change, we basically call
```
for j in 1:length(pairs)
      p = optimize_pair(pairs[j])
end
```
instead of `map(optimize_pair, pairs)`

In either case, we call `optimize_pair` exactly once on each pair. What am I missing?

Reply:
You're right, it's in the nested loop. I don't know how I missed that.